### ISSUE TITLE
Nullable initialization analysis for properties with 'init' accessors

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#44889]
**Version Used**: 5e8c7ad

**Steps to Reproduce**:

```cs
#nullable enable
record Rec {
    string Prop { get; init; }
    private int _x;
    public Rec(int x) => _x = x; // CS8618
}
```

**Expected Behavior**:
Not obvious to me. We could say that since the property is init-only, we instead push the error to any ObjectCreation which does not assign the property in its initializer. However this seems like it would give unexpected warnings in cases where `Rec` actually sets a non-nullable default value for the property internally.

**Actual Behavior**:
`warning CS8618: Non-nullable property 'Prop' is uninitialized. Consider declaring the property as nullable.`

**Comment by jcouv** [comment]
Note: as far as nullability analysis, properties with `init` or `set` accessors behave the same.
Here's an [example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMYHsEBNkA0MIAlgDYG4gDUAPgMQB2cZZUwZEABBI+5wFgAUJhy4uAJQgYuAb2FdFXAAIBGAAxcACgiwAHOVwDmEGAG4uyUxYC+CpXoQkAbrG4lGMLgH0AHmftFZQBmSWkACg8vXwBKLgBeAD4fXwSufy4AekyuAGEAZQAOADZVQuE7IWE+AFsUPSgMbjUAOgk4TxI6ltysGr1yCAR8oecSJuRheSElFVCrKE5xZQAmLgBJZABRXxghvjJ1xhIYQLkzypsgA===) with `set` instead of `init`.

This also relates to discussion to have mandatory properties to initialize (not part of current C# 9 plan).

**Comment by canton7** [comment]
This potentially relates to https://github.com/dotnet/csharplang/issues/2328. One of the suggestions there was to push NRT warnings for `init` properties to the initialization site as well, as a way of avoiding warnings when declaring types meant for deserialization.

**Comment by jcouv** [comment]
@RikkiGibson There seems to be typos in the OP snippet. Can you take a look and clarify?

**Comment by RikkiGibson** [comment]
I didn't see typos in the snippet, and it seems to behave as expected in SharpLab. Maybe we could sync offline and clear up any confusion?

**Comment by jcouv** [comment]
Ah, my bad (I'd misread). 
Confirmed on [sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8DEA7AVwwwmAzgAI49TyBYAKATgGMB7BAEwoCVWKA3owoiKAAQCMABgoAFBGwAOgigHM4MANwU0eNFooBfYaMUI0ANwjwdeGBQD6AD00mRYgMy9WACl32nAEoKAF4APkcnUIoXCgB6OIoAYQBlAA4ANgk0xmMGRhowOChFCBZKSQA6HgI7NCLKpLYwRUxEFMQLNHKoRiEGUXEvSQzxACYKAEkoAFEneAQaDEm9GDdBdbzDIA===)

**Comment by crozone** [comment]
If this is implemented for `init` properties, it should also be implemented for `set` as well for consistency, since conceptually it is the same issue.

The compiler could add per-constructor metadata to properties that are not set by that constructor and then use this to inform which properties are required to be set during construction. However this sounds complicated and a bit *too* opaque and fragile, library authors who are authoring public classes might be prone to accidentally changing the initialization requirements of their classes by simply changing the way properties are set during object construction.

I'm guessing it's easier just to wait for the "mandatory properties to initialize" feature and let the programmer pick and choose which properties must be set. Any uninitialized property that has the "Must be set" flag would not generate a warning at the declaration site, indicating to the class author that either the property must be set, or explicitly declared as a "Must be set" property. It's overall more flexible and more explicitly understandable than automatic nullable initialization analysis from the compiler.

**Comment by RikkiGibson** [comment]
From where I'm standing now, it appears there is no need for analysis of properties with 'init' accessors to differ from analysis of properties with 'set' accessors. The confusion in the issue description is really a request for a "required properties" feature 😉

**Comment by GF-Huang** [comment]
Issue still in vs2019 16.8.4.

**Comment by RikkiGibson** [comment]
See dotnet/csharplang#3630

**Comment by GF-Huang** [comment]
> See [dotnet/csharplang#3630](https://github.com/dotnet/csharplang/issues/3630)

So what's the solution now?

**Comment by RikkiGibson** [comment]
The most straightforward thing to do right now is probably to use positional records (the `record Rec(int x) { }` syntax), or use constructor parameters to assign the properties, or to suppress the initialization warnings.
